Custom and registry cell type documentation (#2334)
astronomer/docs: astro/cloud-ide/custom-cell-reference.md

---
sidebar_label: Custom cell configuration reference
title: Custom cell configuration reference
id: custom-cell-reference

> description: Learn how to configure a custom cell type as JSON.

Author: ```suggestion
description: Learn how to import custom operators by configuring a custom cell type with JSON.
```

---
sidebar_label: Custom cell configuration reference
title: Custom cell configuration reference
id: custom-cell-reference
description: Learn how to configure a custom cell type as JSON.
---

This page contains reference information about all available JSON configurations for the custom operator cell type in the Astro Cloud IDE.  You enter this JSON when you create a new custom operator cell type in the Astro Cloud IDE. See [Create custom operator cells](cloud-ide/use-airflow-operators.md#create-custom-operator-cells) for setup steps.

## Display JSON reference

The following table is a reference for all possible key-value pairs in the `display` section of the JSON request.

> | Name          | Type   | Description                                       |
> | ------------- | ------ | ------------------------------------------------- |
> | `label`       | String | The label to show in the **Add Cell** menu.       |
> | `description` | String | The description to show in the **Add Cell** menu. |

Reviewer: ```suggestion
| Name          | Type   | Description                                       |
| ------------- | ------ | ------------------------------------------------- |
| `label`       | String | The label to show in the **Add Cell** menu.       |
| `description` | String | The description to show in the **Add Cell** menu. |
| `logoUrl` | String | The URL of the logo to show in the **Add Cell** menu. |
```

| `generatesData`    | Boolean                              | Determines whether the cell type generates data for XComs.                                                                                                                         |
| `namingStrategy`   | String, either `increment` or `uuid` | Determines how the default names are generated for new cells.                                                                                                                      |
| `excludeFromGraph` | Boolean                              | Determines whether to hide instances of the cell from the graph view of your pipeline editor.                                                                                      |
| `returnsRawValue`  | Boolean                              | Indicates whether the custom operator returns a raw value. Set to `true` if the operator uses TaskFlow. If set to false, `.output` is appended to the name of the operator output. |

## Generation JSON reference

The following table is a reference for all possible key-value pairs in the `generation` section of the JSON request.

> | Name               | Type                                 | Description                                                                                                                                                                        |
> | ------------------ | ------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
> | `type`         | String, only option is `invoke`.                              | Determines how the operator is called.                                                                                                                                      |
> | `invoke`    | Key-value pairs                              | The configuration for how the operator is invoked. Set `functionName` (String), `imports` (List of Strings), and `excludeTaskArgId` (Boolean)                                                                                                                         |
> | `namingStrategy`   | String, either `increment` or `uuid` | Determines how the default names are generated for new cells.                                                                                                                      |
> | `excludeFromGraph` | Boolean                              | Determines whether to hide instances of the cell from the graph view of your pipeline editor.                                                                                      |
> | `returnsRawValue`  | Boolean                              | Indicates whether the custom operator returns a raw value. Set to `true` if the operator uses TaskFlow. If set to false, `.output` is appended to the name of the operator output. |

Reviewer: ```suggestion
| Name               | Type                                 | Description                                                                                                                                                                        |
| ------------------ | ------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `type`         | String, only option is `invoke`.                              | Determines how the operator is called.                                                                                                                                      |
| `invoke`    | Key-value pairs                              | The configuration for how the operator is invoked. Set `functionName` (String), `imports` (List of Strings), and `excludeTaskArgId` (Boolean)                                                                                                                         |
```

## Configs JSON reference

The following table is a reference for all possible key-value pairs in the `configs` section of the JSON request.

You can configure multiple `configs` objects for each parameter you want the user to enter. 

| Name               | Type                                 | Description                                                                                                                                                                        |
| ------------------ | ------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `key`         | String                             | The unique identifier for the parameter.                                                                                                                                      |

> | `dataType`    | String, either `string`, `stringList`, `boolean`, `integer`                              | The parameter's value type.                                                              |

Reviewer: ```suggestion
| `dataType`    | String, must be one of: `string`, `stringList`, `stringMap`, `stringListMap`, `stringMapList`, `stringSet`, `boolean`, `integer`, `float`, `duration`, `unsupported`                              | The parameter's value type.                                                              |
```

| `validity` | Key-value pairs | Configurations to determine which inputs for the parameter are valid. See [Validity JSON reference](#validity-json-reference)|

### Display JSON reference

For each `configs` object, you can configure the following values in a `display` object to change how the parameter appears to users.

| Name               | Type                                 | Description                                                                                                                                                                        |
| ------------------ | ------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `label`         | String                             | The label for the parameter.                                                                                                                                      |
| `description`    | String                              | The description for the parameter.                                                            |

> | `default`   | Key-value pair. The key becomes the default `dataType` and the value becomes the default value. | The default type and value for the parameter.                                                                                                                      |

Reviewer: we don't actually support this yet
```suggestion
```

| `example` | Key-value pair. The key becomes the default `dataType` and the value becomes the default value.                              | The example value that appears in placeholder text.                                                                                      |

> | `fixedWidthFont`  | Boolean                              | Determines whether to format input as a fixed-width font. |

Reviewer: we don't actually support this yet
```suggestion
```
